lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,202 @@
+        @stderr.close
+
+        exit_msg = "PowerShell process did not terminate in reasonable time"
+        begin
+          Timeout.timeout(3) do
+            Puppet.debug "Awaiting PowerShell process termination..."
+            @exit_status = @ps_process.value
+          end
+        rescue Timeout::Error
+        end
+
+        exit_msg = "PowerShell process exited: #{@exit_status}" if @exit_status
+        Puppet.debug(exit_msg)
+        if @ps_process.alive?
+          Puppet.debug("Forcefully terminating PowerShell process.")
+          Process.kill('KILL', @ps_process[:pid])
+        end
+      end
+
+      def template_path
+        File.expand_path('../../templates', __FILE__)
+      end
+
+      def make_ps_code(powershell_code, output_ready_event_name, timeout_ms = 300 * 1000)
+        template_file = File.new(template_path + "/invoke_ps_command.erb").read
+        template = ERB.new(template_file, nil, '-')
+        template.result(binding)
+      end
+
+      private
+
+      def self.is_readable?(stream, timeout = 0.5)
+        read_ready = IO.select([stream], [], [], timeout)
+        read_ready && stream == read_ready[0][0]
+      end
+
+      def self.create_event(name, manual_reset = false, initial_state = false)
+        handle = FFI::Pointer::NULL_HANDLE
+
+        FFI::Pointer.from_string_to_wide_string(name) do |name_ptr|
+          handle = CreateEventW(FFI::Pointer::NULL,
+            manual_reset ? 1 : FFI::WIN32_FALSE,
+            initial_state ? 1 : FFI::WIN32_FALSE,
+            name_ptr)
+
+          if handle == FFI::Pointer::NULL_HANDLE
+            msg = "Failed to create new event #{name}"
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+        end
+
+        handle
+      end
+
+      WAIT_ABANDONED = 0x00000080
+      WAIT_OBJECT_0 = 0x00000000
+      WAIT_TIMEOUT = 0x00000102
+      WAIT_FAILED = 0xFFFFFFFF
+
+      def self.wait_on(wait_object, timeout_ms = 50)
+        wait_result = Puppet::Util::Windows::Process::WaitForSingleObject(
+          wait_object, timeout_ms)
+
+        case wait_result
+        when WAIT_OBJECT_0
+          Puppet.debug "Wait object signaled"
+        when WAIT_TIMEOUT
+          Puppet.debug "Waited #{timeout_ms} milliseconds..."
+        # only applicable to mutexes - should never happen here
+        when WAIT_ABANDONED
+          msg = 'Catastrophic failure: wait object in inconsistent state'
+          raise Puppet::Util::Windows::Error.new(msg)
+        when WAIT_FAILED
+          msg = 'Catastrophic failure: waiting on object to be signaled'
+          raise Puppet::Util::Windows::Error.new(msg)
+        end
+
+        wait_result
+      end
+
+      def write_stdin(input)
+        @stdin.puts(input)
+      rescue => e
+        msg = "Error writing STDIN / reading STDOUT: #{e}"
+        raise Puppet::Util::Windows::Error.new(msg)
+      end
+
+      def drain_pipe(pipe)
+        output = []
+        while self.class.is_readable?(pipe, 0.1) do
```

Comment:
@glennsarti you are correct that we actually need 2 separate threads draining `stdout` and `stderr` to implement this properly.  Our hokey timeslicing-esque approach is better than nothing, but it's wrong.

This was one of the nice things in DSC about only having to read `stdout` - we were able to dodge some of that complexity.

I think we bound the iterations for now, but flesh out the threaded solution in a subsequent ticket / PR.